File: chapter_8/python/magicIndexDuplicate.py
def magicFast(array):
     return help(array, 0, len(array) - 1)
def help(array, start, end):
     if start > end:
          return -1
     mid = (start + end) / 2
     midValue = array[mid]
     if midValue == mid:
          return mid
     leftEnd = min(mid - 1, midValue)
     left =  help(array, start, leftEnd)
     if left> 0:
          return left
     rightStart = max(mid + 1, midValue)
     right =  help(array, rightStart, end)
     return right
if __name__ == '__main__':
    array = [-10,-5,2,2,2,3,4,7,9,12,13]
    magicIndex = magicFast(array)
    print magicIndex

File: chapter_8/java/MagicIndexDuplicate.java
class MagicIndexDuplicate{
static int magicFast(int[] array){
    return magicFast(array, 0 ,array.length - 1);
}
static int magicFast(int[] array, int start, int end){
    if (start > end){
        return -1;
    }
    int mid = (start + end ) / 2;
    int midValue = array[mid];
    if (midValue == mid){
        return mid;
    }
    int leftEnd = Math.min(mid - 1, midValue);
    int left =  magicFast(array, start, leftEnd);
    if (left > 0){
        return left;
    }
    int rightStart = Math.max(mid + 1, midValue);
    int right = magicFast(array, rightStart, end);
    return right;
    
}
public static void main(String[] args){
 int[] array = {-10,-5,2,2,2,3,4,7,9,12,13};
 int magic_index = magicFast(array);
 System.out.println(magic_index);
}


}
